cxx-qt-lib/include/cxxqtthread.h

```diff
@@ -0,0 +1,70 @@
+// clang-format off
+// SPDX-FileCopyrightText: 2022 Klarälvdalens Datakonsult AB, a KDAB Group company <info@kdab.com>
+// clang-format on
+// SPDX-FileContributor: Andrew Hayzen <andrew.hayzen@kdab.com>
+//
+// SPDX-License-Identifier: MIT OR Apache-2.0
+#pragma once
+
+#include <memory>
+#include <mutex>
+#include <shared_mutex>
+
+#include <QtCore/QDebug>
+#include <QtCore/QMetaObject>
+#include <QtCore/QPointer>
+
+#include "rust/cxx.h"
+
+namespace rust {
+namespace cxxqtlib1 {
+
+template<typename T>
+class CxxQtThread
+{
+public:
+  CxxQtThread(std::shared_ptr<QPointer<T>> ptr,
+              std::shared_ptr<std::shared_mutex> ptrMutex,
+              std::shared_ptr<std::mutex> rustObjMutex)
+    : m_ptr(ptr)
+    , m_ptrMutex(ptrMutex)
+    , m_rustObjMutex(rustObjMutex)
+  {
+  }
+
+  bool queue(rust::Fn<void(T& self)> closure) const
+  {
+    // Ensure that we can read the pointer and it's not being written to
+    const auto guard = std::shared_lock(*m_ptrMutex);
+    if (*m_ptr) {
```

Comment:
We need to do the checks as `QMetaObject::invokeMethod` isn't "safe" to our standards, so we have to check inside the lambda and outside when we call `QMetaObject::invokeMethod` as it could have undefined behaviour inside the `QMetaObject::invokeMethod`.

Reply:
Ah, you mean if the following happens:
1. I request a QtThread instance
2. The QObject is destroyed
3. QtThread::queue is called

I see the problem. Fair Point.